share-coordinator/src/main/java/org/apache/kafka/coordinator/share/ShareCoordinatorService.java

```diff
@@ -240,9 +249,43 @@ public void startup(
 
         log.info("Starting up.");
         numPartitions = shareGroupTopicPartitionCount.getAsInt();
+        setupRecordPruning();
         log.info("Startup complete.");
     }
 
+    private void setupRecordPruning() {
+        timer.add(new TimerTask(config.shareCoordinatorTopicPruneIntervalMs()) {
+            @Override
+            public void run() {
+                for (int i = 0; i < numPartitions; i++) {
```

Comment:
Would It not be the case that each shard should set up the pruning for its owned partitions? Shouldn't the pruning stop when a shard loses leadership of a partition?

Reply:
No - the shard (state machine) does not maintain that mapping. The information is maintained by the runtime which calls the appropriate shards, based on the context. This information is not exposed outside. We need access to https://github.com/apache/kafka/blob/trunk/coordinator-common/src/main/java/org/apache/kafka/coordinator/common/runtime/CoordinatorRuntime.java#L1877 to expose this information. Even then the shard cannot do this.

The runtime is encapsulated in the ShareCoordinatorService and only it can issue calls to the runtime. The Shard only serves to provide data related to partitions.

Using the loop approach - for a specific internal topic-partition only the correct Shard will honour the request and the others will fail silently due to NOT_COORDINATOR.

Flow is 
```
                                                                      ShareCoordinatorShard.callback
                                                                                      |
                                                                                      |  
                                               add task             task with correct shard                     Queue
ShareCoordinatorService ----------->   Runtime -----------> ==================== ----> EventProcessor
                                                                       |
                                                                       |
                                                          obtain shard from TP in task

```

Reply:
Thanks for the explanation.

Reply:
Yes, I've been thinking about this. I'm not entirely comfortable with every broker starting a timer for every partition. I know it's harmless, but it's not exactly elegant.

Reply:
@dajac we don't need the coordinators, but the list of topic partitions whose coordinators are ACTIVE at the point of execution of the timer job.

CoordinatorRuntime
```
    public List<TopicPartition> activeTopicPartitions() {
        return coordinators.entrySet().stream()
            .filter(entry -> entry.getValue().state.equals(CoordinatorState.ACTIVE))
            .map(Map.Entry::getKey)
            .toList();
    }
```


Caller:
```
activeTopicPartitions.forEach(tp -> runtime.scheudle(tp, ...)
```